src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs

```diff
@@ -0,0 +1,203 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers.Binary;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace Nethermind.Serialization.FluentRlp;
+
+public delegate void RefRlpWriterAction<in TContext>(ref RlpWriter arg, TContext value)
+    where TContext : allows ref struct;
+
+public delegate void RefRlpWriterAction(ref RlpWriter arg);
+
+public ref struct RlpWriter
+{
+    private const bool LengthMode = false;
+    private const bool ContentMode = true;
+
+    private bool _mode;
+
+    public int Length { get; private set; }
+
+    private byte[] _buffer;
+    private int _position;
+
+    public static RlpWriter LengthWriter()
+    {
+        return new RlpWriter
+        {
+            _mode = LengthMode
+        };
+    }
+
+    public static RlpWriter ContentWriter(byte[] buffer)
```

Comment:
Have you considered adding an option aligned with `Utf8JsonWriter` where there's a ctor that accepts `IBufferWriter<byte>`? This would mean that ufortunately dicontinued chunks should be supported, but could allow to provide a writer over anything. Maybe this could help to address Netty comment from @LukaszRozmej  

https://learn.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonwriter.-ctor?view=net-9.0#system-text-json-utf8jsonwriter-ctor(system-buffers-ibufferwriter((system-byte))-system-text-json-jsonwriteroptions)

Reply:
Interesting. I picked `byte[]` as a safe default but there is no reason why other type could be used.

- `IBufferWriter<T>` is quite small, it's part of the std lib and it supports `Span`-based APIs.

- `NettyRlpStream` is based on `IByteBuffer`, the latter which is defined in `DotNetty.Buffers`. It supports the operations that we need but it's quite large.

Now, the issue is that `IByteBuffer` and `IBufferWriter` are unrelated so we would need to pick one (unless we start doing conversions).

Reply:
`IBufferWriter` is probably more useful, we can try making adapter to `IByteBuffer` if we want

Reply:
Replaced `byte[]` buffer with `IBufferWriter<byte>`.

The return type of the `Rlp` interface is now `ReadOnlyMemory<byte>` instead of `byte[]` for more flexibility compared to returning `ReadOnlySpan<byte>`. Overloads for `Rlp.Read(ReadOnlyMemory<byte>)` were added.